### ISSUE TITLE
Prototype HyperNetwork for conditioning on Luca's notebook

### ISSUE THREAD

**Issue by mateibejan1** [pinellolab/DNA-Diffusion#44]
Prototype HyperNetwork for weights generation for our main denoising network.

**Comment by Johnny1188** [comment]
Initial experiment with hypernetworks here:
https://github.com/pinellolab/DNA-Diffusion/blob/hypernet-diffusion/notebooks/experiments/conditional_diffusion/hypernet_conditioning.ipynb

Copied from the notebook (short description):

---

**Conditioning the difussion process by hypernetworks**

- This is a prototype of conditioning of the diffusion process by using hypernetworks that generate the parameters for the last blocks of the UNET.
- 90% of the code is adapted from the original notebook on conditional diffusion (Lucas Ferreira da Silva, Luca Pinello, Zach Nussbaum), only with a reorganization of the code and some minor changes.
- As of now, the hypernetworks predict parameters based on the *component id*, not the *cell type*.
- Hypernetworks are trained along with the diffusion process (end-to-end).
- Initial experiments show similar performance (loss; KL divergence) as not using hypernetworks - although over multiple runs, a little faster drop of loss has been observed *(although quite unstable due to very small batch sizes of 14...hitting gpu mem limits)*. Comparisons made only on 4 components, more conditional classes might show a bigger difference using this approach.
- Why hypernetworks:
  - More expressive than classical embedding methods
  - Potential for compression of the parameters (hypernetworks can be seen as a weight-sharing mechanism)
  - Less training data needed for new cell types/components which could be beneficial in the future (reason is the transfer learning of the hypernetwork).
  - Idea: Mixture of denoisers (one hypernetwork generating parameters for multiple denoisers).
  
Contributor:
- Jan Sobotka (github: Johnny1188)

---

Comparison - 1150 epochs:
_Using hypernetwork:_
![unet_hypernet_1150_epochs](https://user-images.githubusercontent.com/50166527/205012939-df0fe40c-ac17-4883-a977-644ff9e04b03.png)

_Not using hypernetwork:_
![unet_no_hypernet_1150_epochs](https://user-images.githubusercontent.com/50166527/205013212-13f589d0-f10a-4d13-99ae-32bad4457302.png)
